File: benchmark/src/ruby/radix_sort.rb
n = 200000
$seed = 12345
def ri(m)
  $seed = ($seed * 16807) % 2147483647
  $seed % m
end

arr = []
n.times { arr << ri(1000000) }

start = Time.now.to_f * 1000

buckets = Array.new(10, 0)
output = Array.new(n)

6.times do |p|
  10.times { |i| buckets[i] = 0 }
  div = 10 ** p
  n.times do |i|
    d = (arr[i] / div) % 10
    buckets[d] += 1
  end
  (1..9).each { |i| buckets[i] += buckets[i - 1] }
  (n - 1).downto(0) do |i|
    d = (arr[i] / div) % 10
    buckets[d] -= 1
    output[buckets[d]] = arr[i]
  end
  arr, output = output, arr
end

c = 0
n.times { |i| c = (c + arr[i] * (i % 7)) % 1000003 }

finish = Time.now.to_f * 1000
puts "Result: #{arr[0]}/#{arr[n-1]}_#{c}"
puts "Time: #{(finish - start).round}ms"

File: benchmarks/src/php/radix_sort.php
<?php
$n = 200000;
$seed = 12345;
function ri($m) {
    global $seed;
    $seed = ($seed * 16807) % 2147483647;
    return $seed % $m;
}

$arr = [];
for ($i = 0; $i < $n; $i++) $arr[] = ri(1000000);

$start = hrtime(true);

$buckets = array_fill(0, 10, 0);
$output = array_fill(0, $n, 0);

for ($p = 0; $p < 6; $p++) {
    for ($i = 0; $i < 10; $i++) $buckets[$i] = 0;
    $div = 10 ** $p;
    for ($i = 0; $i < $n; $i++) {
        $d = intdiv($arr[$i], $div) % 10;
        $buckets[$d]++;
    }
    for ($i = 1; $i < 10; $i++) $buckets[$i] += $buckets[$i - 1];
    for ($i = $n - 1; $i >= 0; $i--) {
        $d = intdiv($arr[$i], $div) % 10;
        $buckets[$d]--;
        $output[$buckets[$d]] = $arr[$i];
    }
    $tmp = $arr; $arr = $output; $output = $tmp;
}

$c = 0;
for ($i = 0; $i < $n; $i++) $c = ($c + $arr[$i] * ($i % 7)) % 1000003;

$end = hrtime(true);
$ms = ($end - $start) / 1e6;
echo "Result: {$arr[0]}/{$arr[$n-1]}_{$c}\n";
echo "Time: " . round($ms) . "ms\n";
